File: RegExr/nestedMatch.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.regex.Pattern;

class nestedMatch {

    // Utility function to create a regex pattern dynamically
    public static String createPattern(int n) {
        return "(a?){" + n + "}(a){" + n + "}";
    }

    // Function to perform regex matching with a timeout
    public static boolean matchWithTimeout(String text, String pattern, long timeout, TimeUnit unit) throws TimeoutException {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Future<Boolean> future = executor.submit(new Callable<Boolean>() {
            @Override
            public Boolean call() {
                Pattern regex = Pattern.compile(pattern);
                return regex.matcher(text).matches();
            }
        });

        try {
            return future.get(timeout, unit);  // Await the result with timeout
        } catch (TimeoutException e) {
            future.cancel(true);  // Cancel the task if timeout occurs
            throw e;
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
            return false;
        } finally {
            executor.shutdown();
        }
    }

    // Main function to run the sequence
    public static void runSequentially(int maxTries, long matchTimeout, TimeUnit unit) {
        List<Integer> nValues = new ArrayList<>();
        List<Double> timeValues = new ArrayList<>();

        for (int n = 1; n <= maxTries; n++) {
            StringBuilder textBuilder = new StringBuilder();
            for (int i = 0; i < n; i++) {
                textBuilder.append("a");
            }
            String text = textBuilder.toString();
            String pattern = createPattern(n);

            long startTime = System.nanoTime();

            try {
                boolean match = matchWithTimeout(text, pattern, matchTimeout, unit);
                long endTime = System.nanoTime();
                double timeTaken = (endTime - startTime) / 1_000_000_000.0;

                if (!match) {
                    System.out.printf("No match for n=%d\n", n);
                    break;
                }

                nValues.add(n);
                timeValues.add(timeTaken);
                System.out.printf("Time taken for n=%d: %.10f seconds\n", n, timeTaken);

            } catch (TimeoutException e) {
                long endTime = System.nanoTime();
                double timeTaken = (endTime - startTime) / 1_000_000_000.0;
                System.out.printf("Match for n=%d timed out after %.10f seconds\n", n, timeTaken);
                break;
            }
        }

        exportToCSV("java-run-stats.csv", nValues, timeValues);
    }

    // Function to export data points to a CSV file
    public static void exportToCSV(String filename, List<Integer> nValues, List<Double> timeValues) {
        try (FileWriter writer = new FileWriter(filename)) {
            writer.append("n,time-seconds\n");
            for (int i = 0; i < nValues.size(); i++) {
                writer.append(nValues.get(i).toString()).append(",")
                        .append(String.format("%.10f", timeValues.get(i))).append("\n");
            }
            System.out.printf("Data successfully exported to %s\n", filename);
        } catch (IOException e) {
            System.out.println("Error creating file: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        long matchTimeout = 10;  // seconds
        int maxTries = 1000;

        runSequentially(maxTries, matchTimeout, TimeUnit.SECONDS);
    }
}

File: RegExr/nestedMatch.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;

// Utility function to create a regex pattern dynamically
string CreatePattern(int n)
{
    return $"(a?){{{n}}}(a){{{n}}}";
}

// Function to perform regex matching with a timeout
async Task<(bool, Exception)> MatchWithTimeoutAsync(string text, string pattern, TimeSpan timeout)
{
    try
    {
        // Run the regex match as a task
        var regexTask = Task.Run(() =>
        {
            Regex regex = new Regex(pattern);
            return regex.IsMatch(text);
        });

        // Wait for either the regex task to complete or the timeout
        if (await Task.WhenAny(regexTask, Task.Delay(timeout)) == regexTask)
        {
            // Regex task completed, return its result
            return (regexTask.Result, null);
        }
        else
        {
            // Timeout
            return (false, new TimeoutException("The operation timed out."));
        }
    }
    catch (Exception ex)
    {
        // Return any exception encountered during the match
        return (false, ex);
    }
}

// Main function to run the sequence of regex matches sequentially, each with a timeout
async Task RunMatchesSequentiallyWithTimeout(int maxTries, TimeSpan matchTimeout)
{
    List<int> nValues = new List<int>();
    List<double> timeValues = new List<double>();

    for (int n = 1; n <= maxTries; n++)
    {
        string text = new string('a', n);
        string pattern = CreatePattern(n);
        DateTime stepStartTime = DateTime.Now;

        (bool match, Exception error) = await MatchWithTimeoutAsync(text, pattern, matchTimeout);
        DateTime stepEndTime = DateTime.Now;
        double timeTaken = (stepEndTime - stepStartTime).TotalSeconds;

        if (error is TimeoutException)
        {
            Console.WriteLine($"Match for n={n} timed out after {timeTaken:F10} seconds");
            break;
        }
        else if (error != null)
        {
            Console.WriteLine($"Error for n={n}: {error.Message}");
            break;
        }

        if (!match)
        {
            Console.WriteLine($"No match for n={n}");
            break;
        }

        nValues.Add(n);
        timeValues.Add(timeTaken);

        Console.WriteLine($"Time taken for n={n}: {timeTaken:F10} seconds");
    }

    ExportToCSV("dotnet-run-stats.csv", nValues, timeValues);
}

// Function to export data points to a CSV file
void ExportToCSV(string filename, List<int> nValues, List<double> timeValues)
{
    using (StreamWriter writer = new StreamWriter(filename))
    {
        writer.WriteLine("n,time-seconds");
        for (int i = 0; i < nValues.Count; i++)
        {
            writer.WriteLine($"{nValues[i]},{timeValues[i]:F10}");
        }
    }

    Console.WriteLine($"Data successfully exported to {filename}");
}

// Entry point of the script
TimeSpan matchTimeout = TimeSpan.FromSeconds(10);
int maxTries = 1000;

// Call the method to run the regex matches sequentially
await RunMatchesSequentiallyWithTimeout(maxTries, matchTimeout);
